engine/src/main/scala/cromwell/services/package.scala

```diff
@@ -0,0 +1,30 @@
+package cromwell
+
+import java.sql.Timestamp
+
+import cromwell.backend.BackendJobDescriptorKey
+import cromwell.core.WorkflowId
+import cromwell.engine.ExecutionIndex._
+
+
+package object services {
+  case class MetadataJobKey(callFqn: String, index: ExecutionIndex, attempt: Int)
+  object MetadataJobKey {
+    def apply(backendJobKey: BackendJobDescriptorKey): MetadataJobKey = MetadataJobKey(backendJobKey.call.fullyQualifiedName, backendJobKey.index, backendJobKey.attempt)
+  }
+
+  case class MetadataKey(workflowId: WorkflowId, jobKey: Option[MetadataJobKey], key: String)
+  case class MetadataValue(value: String)
+  case class MetadataEvent(key: MetadataKey, value: MetadataValue, timestamp: Timestamp)
+
+  case class MetadataQueryJobKey(callFqn: String, index: ExecutionIndex, attempt: Int)
+  object MetadataQueryJobKey {
+    def forMetadataJobKey(jobKey: MetadataJobKey) = MetadataQueryJobKey(jobKey.callFqn, jobKey.index, jobKey.attempt)
+  }
+  case class MetadataQuery(workflowId: WorkflowId, jobKey: Option[MetadataQueryJobKey], key: Option[String])
```

Comment:
Note!  I changed MetadataQuery from having a parameter `Option[WorkflowId]` to just `WorkflowId` since I didn't think there was a legitimate use case for a workflow ID-less query.  All the MetadataBuilder* changes here are a consequence of this.